wdl-analysis/src/config.rs

```diff
@@ -0,0 +1,246 @@
+// arc and the private inner struct
+impl std::fmt::Debug for Config {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Config")
+            .field("diagnostics_config", &self.inner.diagnostics_config)
+            .field("fallback_version", &self.inner.fallback_version)
+            .finish()
+    }
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            inner: Arc::new(ConfigInner {
+                diagnostics_config: Default::default(),
+                fallback_version: None,
+            }),
+        }
+    }
+}
+
+impl Config {
+    /// Get this configuration's [`DiagnosticsConfig`].
+    pub fn diagnostics_config(&self) -> &DiagnosticsConfig {
+        &self.inner.diagnostics_config
+    }
+
+    /// Get this configuration's fallback version; see
+    /// [`Config::with_fallback_version()`].
+    pub fn fallback_version(&self) -> Option<SupportedVersion> {
+        self.inner.fallback_version
+    }
+
+    /// Return a new configuration with the previous [`DiagnosticsConfig`]
+    /// replaced by the argument.
+    pub fn with_diagnostics_config(&self, diagnostics: DiagnosticsConfig) -> Self {
+        let mut inner = (*self.inner).clone();
+        inner.diagnostics_config = diagnostics;
+        Self {
+            inner: Arc::new(inner),
+        }
+    }
+
+    /// Return a new configuration with the previous version fallback option
+    /// replaced by the argument.
+    ///
+    /// This option controls what happens when analyzing a WDL document with a
+    /// syntactically valid but unrecognized version in the version
+    /// statement. The default value is `None`, with no fallback behavior.
+    ///
+    /// Configured with `Some(fallback_version)`, analysis will proceed as
+    /// normal if the version statement contains a recognized version. If
+    /// the version is unrecognized, analysis will continue as if the
+    /// version statement contained `fallback_version`, though the concrete
+    /// syntax of the version statement will remain unchanged.
+    ///
+    /// <div class="warning">
+    ///
+    /// # Warnings
+    ///
+    /// This option is intended only for situations where unexpected behavior
+    /// due to unsupported syntax is acceptable, such as when providing
+    /// best-effort editor hints via `wdl-lsp`. The semantics of executing a
+    /// WDL workflow with an unrecognized version is undefined and not
+    /// recommended.
+    ///
+    /// Once this option has been configured for an `Analyzer`, it should not be
+    /// changed. A document that was initially parsed and analyzed with one
+    /// fallback option may cause errors if subsequent operations are
+    /// performed with a different fallback option.
+    ///
+    /// </div>
+    pub fn with_fallback_version(&self, fallback_version: Option<SupportedVersion>) -> Self {
+        let mut inner = (*self.inner).clone();
```

Comment:
Just curious, will this `Arc::clone` get optimized out if you call `Config::default().with_fallback_version(...)`?

Reply:
If this is desired, we could take `mut self` and do a `Arc::make_mut` on inner, saving a clone + allocation when there is only one strong ref. It's a pretty minor thing, though.

Reply:
The performance impact of a single allocation is vanishingly small, so I'd prefer to prioritize ease of use and maintainability. In particular, the [place where we use `excepted_for_node()`](https://github.com/stjude-rust-labs/wdl/pull/475/files#diff-ed876068b9d45d42a05374ad0b11d06cd51e946ed0b35f02953a50282d74adb7R549-R551) to modify the diagnostics for the version statement does need to keep the original `self`, and making that work with an entirely separate builder type (vs these methods which are a kind of builder-in-place) would make this code more complicated.

Reply:
Apologies, this is another case where I was inadvertently responding to a stale view of the thread.

> If this is desired, we could take `mut self` and do a `Arc::make_mut` on inner, saving a clone + allocation when there is only one strong ref. It's a pretty minor thing, though.

I was unaware of `make_mut`, that sounds like a fine way to do this, especially since it can reduce the clone of the `ConfigInner`.